lona/html/data_binding/select.py

```diff
@@ -80,73 +154,109 @@ def readonly(self, new_value):
+            if not isinstance(new_options, (list, tuple)):
+                new_options = [new_options]
 
-                self.append(option_node)
+            for option in new_options:
+                self.add_option(option)
 
-    # value
+    # selected options
     @property
-    def value(self):
+    def selected_options(self):
         with self.lock:
-            value = []
+            options = ()
+
+            for option in self.options:
+                if not option.selected:
+                    continue
 
-            for option in self.nodes:
-                if 'selected' in option.attributes:
-                    value.append(option.attributes['value'])
+                options += (option, )
 
-            if not self.multiple and not value and self.nodes:
-                option = self.nodes[0]
+            if not options and self.options and not self.multiple:
+                return (self.options[0], )
 
-                value.append(option.attributes['value'])
+            return options
 
-            if not value:
-                if not self.multiple:
-                    return None
+    # value
+    @property
+    def value(self):
+        with self.lock:
+            values = ()
 
-                return value
+            for option in self.selected_options:
+                values += (option.value, )
 
             if not self.multiple:
-                value = value.pop()
+                if not values:
+                    return self.options[0].value
 
-            return value
+                return values[-1]
+
+            return values
 
     @value.setter
     def value(self, new_value):
-        if not isinstance(new_value, list):
-            new_value = [new_value]
+        with self.lock:
+            if not isinstance(new_value, list):
+                new_value = [new_value]
+
+            old_values = self.values
+
+            for value in new_value:
+                if value not in old_values:
+                    raise RuntimeError(f'unknown value: {value}')
 
+            for option in self.options:
+                option._set_selected(option.value in new_value)
+
+    # values
+    @property
+    def values(self):
         with self.lock:
-            for option in self.nodes:
-                if option.attributes['value'] in new_value:
-                    option.attributes['selected'] = ''
+            values = ()
+
+            for option in self.options:
+                values += (option.value, )
 
-                else:
-                    del option.attributes['selected']
+            return values
+
+    # helper ##################################################################
+    def add_option(self, option):
+        with self.lock:
+            option._select = self
+            self.nodes.append(option)
+
+    def remove_option(self, option):
+        with self.lock:
+            option._select = None
+            self.nodes.remove(option)
+
+    def select_all(self):
```

Comment:
Maybe need to throw exception for non-multi select?

Reply:
Ack, Fixed!